OpenRA.Mods.Common/Traits/ExitsDebugOverlay.cs

```diff
@@ -0,0 +1,120 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2016 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Drawing;
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common.Commands;
+using OpenRA.Mods.Common.Graphics;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Displays `Exit` data for factories.")]
+	public class ExitsDebugOverlayInfo : ITraitInfo, Requires<ExitInfo>, Requires<BuildingInfo>
+	{
+		[Desc("The font used to draw exit cell vectors. Should match the value as-is in the Fonts section of the mod manifest (do not convert to lowercase).")]
+		public readonly string Font = "TinyBold";
+
+		[Desc("Should cell vectors be drawn for each perimeter cell?")]
+		public readonly bool DrawCellVectors = true;
+
+		[Desc("Should lines be drawn for each exit?")]
+		public readonly bool DrawSpawnOffsetLines = true;
+
+		public object Create(ActorInitializer init) { return new ExitsDebugOverlay(init.Self, this); }
+	}
+
+	public class ExitsDebugOverlay : INotifyCreated, IChatCommand, IPostRender
+	{
+		const string CommandName = "exits-overlay";
+		const string CommandHelp = "Displays exits for factories.";
+
+		readonly ExitsDebugOverlayInfo info;
+		readonly SpriteFont font;
+		readonly RgbaColorRenderer rgbaRenderer;
+
+		CPos[] exitCells;
+		WPos[] spawnPositions;
+		CPos[] perimeterCells;
+
+		public bool Enabled;
+
+		public ExitsDebugOverlay(Actor self, ExitsDebugOverlayInfo info)
+		{
+			this.info = info;
+
+			if (!Game.Renderer.Fonts.TryGetValue(info.Font, out font))
+				throw new YamlException("Could not find font '{0}'".F(info.Font));
+
+			rgbaRenderer = Game.Renderer.WorldRgbaColorRenderer;
+		}
+
+		void INotifyCreated.Created(Actor self)
+		{
+			var exits = self.Info.TraitInfos<ExitInfo>();
+
+			exitCells = exits.Select(e => self.Location + e.ExitCell).ToArray();
+			spawnPositions = exits.Select(e => self.CenterPosition + e.SpawnOffset).ToArray();
```

Comment:
I suppose I could calculate perimeter cells based on the existence of Building (or if it doesn't exist I can just expand `self.Location`).